src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs

```diff
@@ -2322,67 +2328,259 @@ internal override Binder GetBinder(SyntaxNode node)
+                }
+
+                return block;
+            }
+        }
+
+#nullable enable
+
+        /// <summary>
+        /// This class implements a weak cache of bound nodes from a simple program method body.
+        /// Only bound nodes that can be reused by <see cref="IncrementalBinder"/> are supposed to be placed in this cache.
+        /// The cache is used to cache bound nodes across all compilation units with top-level statements. This
+        /// allows us to avoid re-binding over and over again (assuming nodes in the cache are still alive) compilation units
+        /// that are not associated with specific SemanticModel instance. Binding those compilation units is required
+        /// for getting complete simple program body. To do a flow analysis, for example.
+        /// 
+        /// The cache doesn't keep strong references to the syntax nodes used as keys and bound nodes used as values.
+        /// So, both can be collected when no one else uses them.
+        /// 
+        /// <see cref="SynthesizedSimpleProgramEntryPointSymbol"/> is responsible for creating this cache and sharing it
+        /// between different <see cref="MethodBodySemanticModel"/> instances.
+        /// </summary>
+        internal class SimpleProgramBodySemanticModelMergedBoundNodeCache
+        {
+            /// <summary>
+            /// A helper class that allows us to avoid exposing <see cref="_nodeMap"/> outside and 
+            /// simplifies storage of bound nodes.
+            /// </summary>
+            private class WeakReferenceWrapper
+            {
+                public WeakReference<BoundNode>? WeakReference;
+            }
+
+            /// <summary>
+            /// Different instances of <see cref="MethodBodySemanticModel"/> can use the cache as long as they are based  
+            /// on the same instance of <see cref="SimpleProgramBinder"/>. Otherwise, symbols for locals and labels won't match. 
+            /// We store a weak reference to the binder with the cache so that we can easily determine whether its
+            /// usage is appropriate.
+            /// </summary>
+            public readonly WeakReference<SimpleProgramBinder> WeakBodyBinder;
+
+            private readonly ConditionalWeakTable<SyntaxNode, WeakReferenceWrapper> _nodeMap = new ConditionalWeakTable<SyntaxNode, WeakReferenceWrapper>();
+
+            public SimpleProgramBodySemanticModelMergedBoundNodeCache(SimpleProgramBinder binder)
+            {
+                WeakBodyBinder = new WeakReference<SimpleProgramBinder>(binder);
+            }
+
+            /// <summary>
+            /// Returns a reference to a weak reference stored in the cache. The reference can be used to get/update the cached bound node.
+            /// </summary>
+            public ref WeakReference<BoundNode>? GetValue(SyntaxNode key)
+            {
+                return ref _nodeMap.GetValue(key, k => new WeakReferenceWrapper()).WeakReference;
```

Comment:
>GetValue(key, k => new WeakReferenceWrapper()) [](start = 36, length = 46)

`GetOrCreateValue()`

Reply:
> GetOrCreateValue()

That is probably using Reflection to insatiate values. I prefer this approach. 

---
In reply to: [384776410](https://github.com/dotnet/roslyn/pull/41926#discussion_r384776410) [](ancestors = 384776410)